8273743: KeyCharacterCombination for "+" does not work on US QWERTY keyboard layout

beldenfox opened the pull request:
The algorithm in `KeyCharacterCombination.match` relies on the call `Toolkit.getKeyCodeForChar` which is difficult to implement correctly. It defies the way most keyboard API’s work and no platform has got it right yet. In particular the Mac and Linux implementations have to resort to a brute-force approach which monitors keystrokes to learn the relationship between keys and characters.

This PR introduces an alternative mechanism which directly asks the platform whether a given key can generate a specific character. It also allows the platform to attach identifying key information to each KeyEvent to make it easier to answer the question (much, much easier).

This is mostly dumb plumbing. On the front-end there’s a new call `View.notifyKeyEx` that takes an additional platform-specific `hardwareCode` parameter. It also returns a boolean indicating whether the event was consumed or not so I can fix JDK-8087863. If you want to follow the path visit the files in this order:

	View.java
	GlassViewEventHandler.java
	TKSceneListener.java
	Scene.java

The `KeyEvent` class has been expanded with an additional `hardwareCode` member that can only be accessed internally. See KeyEvent.java and KeyEventHelper.java.

On the back-end `KeyCharacterCombination.match` calls a new routine `Toolkit.getKeyCanGenerateCharacter` which unpacks the `KeyEvent` information and sends it on to the Application. The default implementation falls back to the old `getKeyCodeForChar` call but platform specific Applications can send it on to the native glass code.

	KeyCharacterCombination.java
	Toolkit.java
	QuantumToolkit.java
	Application.java
	GtkApplication.java

The glass code can use the `hardwareCode` to answer the question directly. It also has enough information to fall back on the old `getKeyCodeForChar` logic while also enabling the keypad (a common complaint is that Ctrl+’+’ only works on the main keyboard and not the keypad, see JDK-8090275).

This PR improves the situation for key events generated by keystrokes. Manually constructed key events won’t work any better or worse than they already do. Based on the bug database I don't think this is an issue.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8273743](https://bugs.openjdk.org/browse/JDK-8273743): KeyCharacterCombination for "+" does not work on US QWERTY keyboard layout (**Bug** - P3)


### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**) 🔄 Re-review required (review applies to [ddaf808c](https://git.openjdk.org/jfx/pull/694/files/ddaf808c56ba51a76fda160844bdfcb56ca7ea0c))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/694/head:pull/694` \
`$ git checkout pull/694`

Update a local copy of the PR: \
`$ git checkout pull/694` \
`$ git pull https://git.openjdk.org/jfx.git pull/694/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 694`

View PR using the GUI difftool: \
`$ git pr show -t 694`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/694.diff">https://git.openjdk.org/jfx/pull/694.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/694#issuecomment-992824268)

kevinrushforth commented:
/reviewers 2

kevinrushforth commented:
I see from your comment in PR #635 that you propose to withdraw that PR in favor of this one. Given that, why don't you change that one to "Draft" and we'll focus on this one. As a first pass, it does seem a reasonable approach.

I do have a few high level questions:

1. What are the drawbacks of this approach?
2. I see that the new mechanism has a Linux implementation. Do you plan to propose this for the other platforms (particularly macOS)? It seems like you would need to in order to solve [JDK-8090275](https://bugs.openjdk.java.net/browse/JDK-8090275)
3. Would this PR help address any of the other in-flight PRs? (e.g., PR #672) presuming we eventually do this on all platforms)?
4. Can you add the manual test from PR #635 ?

beldenfox commented:
1. No drawbacks that I can think of. With this approach the platform glass code will have enough information to recreate the original `getKeyCodeForChar` algorithm so it can at least maintain the existing behavior. This PR just gives it enough information to do better. For example, even in the absence of a `hardwareCode` the platform will be able to use the event's `KeyCode` to allow matching on keypad keys.
2. I've written and tested Windows and Mac implementations. They both address [JDK-8090275](https://bugs.openjdk.java.net/browse/JDK-8090275). The Mac implementation requires more testing and uses some code from PR #425. The Windows implementation is just a tweak on the existing code to enable the keypad.
3. I'm not sure what to say about #672. It boils down to how we want to deal with `KeyEvents` that are created using the API vs. a physical keystroke. I'll address that in a follow-up comment.
4. I'll add the test case modified to enable testing on the keypad. It might take a day or two, there's a few other tweaks that would be useful to have.

beldenfox commented:
I've decided that if this PR goes forward we should abandon #672. The changes in #672 would be bypassed for KeyEvents generated by physical keystrokes. For a while I thought it would be useful for handling KeyEvents created via the API but have decided against it.

I've been going back and forth on how to handle synthetic events created using the KeyEvent constructor i.e. events that don't have a hardwareCode attached. One option is to just route them through the old `getKeyCodeForChar` machinery to maintain the current level of support. Unfortunately this varies widely, from mostly working on Windows to almost never working on Mac. If we just maintain what we've got I would ditch #672 since it would be fine-tuning a feature on Windows that's largely busted on other platforms.

The other option is to dive in and make synthetic events work as well as possible on every platform. In this PR `getKeyCanGenerateCharacter` has a KeyCode in hand and can use the existing Robot code to map that to a physical key. It wouldn't be entirely reliable (particularly on Mac and Linux) but it would be easy and improve support for synthetic events enough to forestall bugs like [JDK-8087486](https://bugs.openjdk.java.net/browse/JDK-8087486). And it would be behavior I could write automated and manual tests for.

I don't think synthetic events are all that useful for triggering KeyCharacterCombinations (you have to know the current keyboard layout to accurately target punctuation or symbols) but they exist and it's probably best to get them to a testable state. This means fixing the Robot code which has bugs on every platform but that was already on my to-do list and I've prototyped most of the code. I'll get the Robot bugs entered in the next few days.

kevinrushforth commented:
> I've decided that if this PR goes forward we should abandon #672. The changes in #672 would be bypassed for KeyEvents generated by physical keystrokes. For a while I thought it would be useful for handling KeyEvents created via the API but have decided against it.

OK, I'll focus on this one then, along with #425.

kevinrushforth commented:
@andy-goryachev-oracle or @jperedadnr would one of you also be able to review this?

andy-goryachev-oracle commented:
@beldenfox would it be possible to make the same change to the .classpath file (though it might create a merge conflict later)?

```
<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry combineaccessrules="false" kind="src" path="/base">
		<attributes>
			<attribute name="module" value="true"/>
		</attributes>
	</classpathentry>
	<classpathentry combineaccessrules="false" kind="src" path="/controls">
		<attributes>
			<attribute name="module" value="true"/>
		</attributes>
	</classpathentry>
	<classpathentry combineaccessrules="false" kind="src" path="/graphics">
		<attributes>
			<attribute name="module" value="true"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER">
		<attributes>
			<attribute name="module" value="true"/>
		</attributes>
	</classpathentry>
	<classpathentry excluding=".classpath|.project|.settings" kind="src" output="bin" path=""/>
	<classpathentry kind="output" path="bin"/>
</classpath>

```

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/TKSceneListener.java:
```diff
@@ -64,7 +64,7 @@ public void mouseEvent(EventType<MouseEvent> type, double x, double y, double sc
      * Pass a key event to the scene to handle
      */
-    public void keyEvent(KeyEvent keyEvent);
+    public boolean keyEvent(KeyEvent keyEvent);
```
could we update javadoc to explain when it returns true?

andy-goryachev-oracle commented:
The ticket is a bit unclear when it talks about "US Layout" and Ctrl-+.

For example, on a Mac with an attached USB 101(?) key IBM keyboard, one can find two + keys.  One +/=, and one on the key pad (the latter generates "Ignored: keypad code" - is this expected?)

andy-goryachev-oracle commented:
On Mac, Ctrl-_/- generates "Ignored: control key" - is this expected?

andy-goryachev-oracle commented:
For some reason, Command-+ on Mac did not generate console output in KeyCharComboTest (in the ticket).
Using Windows+L (on the attached IBM keyboard) did generate stdout, but not Windows-+.

Am I doing something wrong?

beldenfox commented:
@andy-goryachev-oracle This bug is Linux-specific and this PR only fixes Linux. It's still worth testing on Mac and Windows just to verify that they're not affected. Those platforms also have problems with KeyCharacterCombinations but I entered separate bugs, see [JDK-8274967](https://bugs.openjdk.org/browse/JDK-8274967) for Windows and [JDK-8087700 ](https://bugs.openjdk.org/browse/JDK-8087700) for Mac. I updated my comment in the Mac bug to better explain the behavior you're seeing.

Since this PR was submitted I have written the KeyboardTest app submitted as part of PR #425 which includes tests for KeyCharacterCombinations and covers a lot of keys in one go. Unfortunately that test relies on robust Robot code and right now only Windows has that (Linux is waiting on #718 and Mac on #425). It's also good to have the final Robot code in place since it's reasonable for an implementation to re-use that code when transitioning to `Toolkit.getKeyCanGenerateCharacter`. 

The core Java code in the PR is mostly plumbing and is still worth reviewing. I needed to transition at least one platform to the new machinery to validate it but in retrospect Linux was probably not the right choice. Eventually I should close out this PR and re-submit the code with the Windows front- and back-ends in place so we can use the semi-automated KeyboardTest app instead.

BTW, when I originally started working on this I wasn't planning on getting KeyCharacterCombinations working with the numeric keypad. That's why the original bug report assumed all testing would be restricted to the main keyboard and the test app in this PR prevents testing on the keypad. But the new Toolkit call makes it (relatively) easy to fix [JDK-8090275](https://bugs.openjdk.org/browse/JDK-8090275) so all that's obsolete. Sorry for the confusion.

andy-goryachev-oracle commented:
@beldenfox thank you for doing this work!  Should we be reviewing/integrating PRs in a specific sequence?

beldenfox commented:
@andy-goryachev-oracle My highest priorities are #425 (Mac) and the Java code in this PR. I'll get it switched over to Windows as soon as possible so we have a complete implementation of the new key event pipeline to review. Third on the list is the Linux Robot code (#718) but the documentation of that API is, um, scattered so I don't expect the review to go all that quickly.

andy-goryachev-oracle commented:
Thank you for clarifications, @beldenfox .

andy-goryachev-oracle reviewed (approved):
looks good to me, as far as I can tell.
needs a second reviewer familiar with the native linux code.

tsayao commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/TKSceneListener.java:
```diff
@@ -63,8 +63,11 @@ public void mouseEvent(EventType<MouseEvent> type, double x, double y, double sc
      * Pass a key event to the scene to handle
+     *
+     * @param keyEvent The key event
+     * @return true iff the event was consumed
```
There's an extra f in "if".

tsayao commented on modules/javafx.graphics/src/main/java/com/sun/glass/ui/View.java:
```diff
@@ -963,7 +968,12 @@ protected void notifyScroll(int x, int y, int xAbs, int yAbs,
+        handleKeyEvent(System.nanoTime(), type, keyCode, keyChars, modifiers, -1);
+    }
+
+    // Returns true iff event was consumed
```
There are some places where it's written "iff", I'm not sure if it's intentional or a typo.

beldenfox replied:
iff is an abbreviation for "if and only if". It's a bit formal but I noticed it was used in other internal API comments and JavaDocs. I think I removed all of these when I resubmitted this code as part of #1126.

tsayao commented on tests/manual/events/KeyCharacterCombinationTest.java:
```diff
@@ -0,0 +1,141 @@
+    private KeyEvent lastPressed = null;
+
+    public static void main(String[] args) {
+        Application.launch(args);
```
Would be nice to add the class to  `launch`, `Application.launch(KeyCharacterCombinationTest.class, args);` so it can be run as:

`java @build/run.args tests/manual/events/KeyCharacterCombinationTest.java`

tsayao commented on tests/manual/events/KeyCharacterCombinationTest.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Override
+    public void start(Stage stage) {
```
I started the test app but was unsure how to test it. Maybe add a header text that explains how a successful test should go.

beldenfox replied:
There's now a better test app in tests/manual/events/KeyboardTest.java. With KeyCharacterCombinationTest the human tester has to press a bunch of keys. KeyboardTest.java uses a Robot to automate the process. The downside is that KeyboardTest.java relies on working Robot code so it can't really be used on Linux until #718 is integrated. I'll update the comments on this in the interim.

beldenfox commented:
@andy-goryachev-oracle My apologies, I really should have cleaned this PR up weeks ago when the bridgekeeper bot flagged it.

The portion of this PR that is platform-agnostic was also submitted in PR #1126. In that PR I incorporated all of the feedback you already provided (and thank you for that). I re-submitted all of this with a Windows front- and back-end because at the time Windows was the only platform with working Robot code so the changes could be tested (the manual app is included in #1126 as well as #425). I was also hoping that the review would go faster on Windows since no one seems to have much familiarity with the Linux API's except me and @tsayao (and the documentation is wretched).

I've already done a bit of cleanup over in #1126 that conflict with the changes here (I just renamed some routines). I've kept this PR around so the Linux-specific code can be reviewed but I should probably switch it over to Draft. And the Linux Robot code needs to be reviewed and integrated first to enable testing (#718).

tsayao commented:
@andy-goryachev-oracle Linux code looks correct to me.

andy-goryachev-oracle commented:
> The portion of this PR that is platform-agnostic was also submitted in PR #1126.

will take a look, thanks!
